app/api/csv/arrangeThesauri.ts

```diff
@@ -0,0 +1,180 @@
+          nameToThesauriIdSelects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      } else if (p.type === propertyTypes.multiselect) {
+        nameToThesauriIdMultiselects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdMultiselects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      }
+    }
+  });
+
+  return [nameToThesauriIdSelects, nameToThesauriIdMultiselects];
+};
+
+type ThesauriValueData = {
+  thesauriIdToExistingValues: Map<string, Set<string>>;
+  thesauriIdToNewValues: Map<string, Set<string>>;
+  thesauriIdToNormalizedNewValues: Map<string, Set<string>>;
+};
+
+const setupIdValueMaps = (allRelatedThesauri: ThesaurusSchema[]): ThesauriValueData => {
+  const thesauriIdToExistingValues = new Map();
+  const thesauriIdToNewValues = new Map();
+  const thesauriIdToNormalizedNewValues = new Map();
+
+  allRelatedThesauri.forEach(t => {
+    if (t._id) {
+      const id = t._id.toString();
+      thesauriIdToExistingValues.set(
+        id,
+        new Set(t.values?.map(v => normalizeThesaurusLabel(v.label)))
+      );
+      thesauriIdToNewValues.set(id, new Set());
+      thesauriIdToNormalizedNewValues.set(id, new Set());
+    }
+  });
+
+  return { thesauriIdToExistingValues, thesauriIdToNewValues, thesauriIdToNormalizedNewValues };
+};
+
+const handleLabels = (
+  id: string,
+  original: string,
+  normalized: string | null,
+  thesauriValueData: ThesauriValueData
+) => {
+  if (
+    normalized &&
+    !thesauriValueData.thesauriIdToExistingValues.get(id)?.has(normalized) &&
+    !thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.has(normalized)
+  ) {
+    thesauriValueData.thesauriIdToNewValues.get(id)?.add(original);
+    thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.add(normalized);
+  }
+};
+
+const syncSaveThesauri = async (
+  allRelatedThesauri: ThesaurusSchema[],
+  thesauriIdToNewValues: Map<string, Set<string>>
+) => {
+  for (let i = 0; i < allRelatedThesauri.length; i += 1) {
+    const thesaurus = allRelatedThesauri[i];
+    if (thesaurus?._id) {
+      const newValues: { label: string }[] = Array.from(
+        thesauriIdToNewValues.get(thesaurus._id.toString()) || []
+      ).map(tval => ({ label: tval }));
+      if (newValues.length > 0) {
+        const thesaurusValues = thesaurus.values || [];
+        // eslint-disable-next-line no-await-in-loop
+        await thesauri.save({
+          ...thesaurus,
+          values: thesaurusValues.concat(newValues),
+        });
+      }
+    }
+  }
+};
+
+const arrangeThesauri = async (
+  file: ImportFile,
+  template: TemplateSchema,
+  languages?: string[],
+  stopOnError: boolean = true
+) => {
+  const thesauriRelatedProperties = template.properties?.filter(p =>
+    ['select', 'multiselect'].includes(p.type)
+  );
+
+  let [
+    nameToThesauriIdSelects,
+    nameToThesauriIdMultiselects,
+  ] = separateSelectAndMultiselectThesauri(thesauriRelatedProperties, languages);
```

Comment:
@LaszloKecskes do you think we can avoid this separation and treat every property as the "same" this i think is possible and removes a lot of code.
- everything is a thesauri, it does not matter if its multi or not, you can parse it assuming always multiple values.

Reply:
This is one of those "why didn't I think of that" moments. Of course you can do that.